### ISSUE TITLE
Encoded RPC Message bytes are not same

### ISSUE THREAD

**Issue by armaganyildirak** [sigp/discv5#207]
When RPC Message was [decoded](https://github.com/sigp/discv5/blob/a9f1e99321aec746fb9d6e8df889aa515a5e1254/src/rpc.rs#L304) and [encoded](https://github.com/sigp/discv5/blob/a9f1e99321aec746fb9d6e8df889aa515a5e1254/src/rpc.rs#L297) these bytes ```[6, 193, 0, 75, 252]```,  Encoding message returns different result ```[6, 194, 0, 75]```

When the message is decoded, The message is ```Response(Response { id: RequestId([0]), body: Talk { response: [75] } })```

**Comment by divagant-martian** [comment]
I did a bit of digging,

first I checked the spec against the code and at least to my eyes, encoding and decoding looks legit. Maybe it could be polished but it seems functionally correct to me.

Then I also did this test
```rust
        let data = [6, 193, 0, 75, 252];
        let msg = Message::decode(&data).unwrap();

        let data2 = [6, 194, 0, 75];
        let msg2 = Message::decode(&data2).unwrap();

        assert_eq!(msg, msg2);
```

It seems to be both representations produce the same message, so I'm no longer sure if this is an issue. It depends on rlp producing unique representations. And I'm not sure that's true

**Comment by weipin** [comment]
[ Looked at this VERY briefly -- there are things I could be missing ]

There could be bugs -- either the RLP crate is misused, or the RLP crate 
itself fails to report certain errors.

Two issues: A) mismatch of "the number of list items" isn't reported. B) trailing
"garbage" of list items isn't reported.

Extending @divagant-martian 's code a little bit (added `data3`):

```
    let data = [6, 193, 0, 75, 252];
    let msg = Message::decode(&data).unwrap();

    let data2 = [6, 194, 0, 75];
    let msg2 = Message::decode(&data2).unwrap();

    let data3 = [6, 194, 0, 75, 252];
    let msg3 = Message::decode(&data3).unwrap();

    assert_eq!(msg, msg2);
    assert_eq!(msg, msg3);
```

`data` demonstrates issue A, `data3` demonstrates issue B. Both should fail and
return proper errors.

Here is a breakdown of decoding `data2`, the seemingly correct one.

Decoding `[6, 194, 0, 75]`:
  * `6`: message type for TALKRESP Response (0x06)
  * `194`: RLP list header, represents 2 items following (2 = 194 - 192) (see RLP spec 1 at the bottom)
  * `0`: the first item, represents byte `0` (request-id) -- see RLP spec 2
  * `75`: the second item, represents byte `75`(response) -- see RLP spec 2

Hope this could make sense.

Related RLP spec
----
RLP spec 1: "If the total payload of a list (i.e. the combined length of all its items being RLP encoded) is 0-55 bytes long, the RLP encoding consists of a single byte with value 0xc0 plus the length of the list followed by the concatenation of the RLP encodings of the items. The range of the first byte is thus [0xc0, 0xf7] (dec. [192, 247])."

RLP spec 2: "For a single byte whose value is in the [0x00, 0x7f] (decimal [0, 127]) range, that byte is its own RLP encoding."

**Comment by divagant-martian** [comment]
I checked with the geth impl (`p2p/discover`) and confirmed only the second one is rejected
for completeness
```rust
	msg := []byte{6, 194, 0, 75}
	p, err := v5wire.DecodeMessage(msg[0], msg[1:])
	if err != nil {
		t.Fail()
	}
	fmt.Printf("%v\n", p)
```
is accepted, and it produces the desired response.
The first one produces an error

**Comment by weipin** [comment]
Tend to agree that the crate doesn't seem to provide the necessary invariant.

Took the liberty of creating another PR for ENR, to remind the same issue. Also searched the CL repos, seem fine. I guess RLP isn't much used there.

**Comment by divagant-martian** [comment]
the mentioned test cases were fixed in #208 
if new failing test cases surface let's create a new isue
